docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,232 @@
+is thus a generalization of the current `map[string]string` but still serves the
+purpose of restricting the set of nodes onto which the Pod can schedule
+* a field called `SoftNodeAffinity` that specifies which nodes are
+preferred for scheduling among those that meet the hard requirements described
+by `HardNodeAffinity`.
+(In practice, as discussed later, we will actually *add* the `Affinity` field
+rather than replacing `map[string]string`, due to backward compatibility requirements.)
+
+`HardNodeAffinity` and `SoftNodeAffinity` allow a pod to request various properties
+that are inherent to nodes, for example "run this pod on a node with an Intel CPU" or, in a
+multi-zone cluster, "run this pod on a node in zone Z."
+([This issue](https://github.com/kubernetes/kubernetes/issues/9044) describes
+some of the properties that a node might publish as labels, which `HardNodeAffinity`
+and `SoftNodeAffinity` can match against.)
+They do *not* allow a pod to request to schedule
+(or not schedule) on a node based on what other pods are running on the node. That
+feature is called "inter-pod topological affinity/anti-afinity" and is described
+[here](https://github.com/kubernetes/kubernetes/pull/18265).
+
+## API
+
+### NodeSelector
+
+```go
+// A node selector represents the union of the results of one or more label queries
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms. An empty node selector matches all objects. A null
+// node selector matches no objects.
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
+	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms,omitempty"`
+}
+
+// An empty node selector term matches all objects. A null node selector term
+// matches no objects.
+type NodeSelectorTerm struct {
+	// matchExpressions is a list of node selector requirements. The requirements are ANDed.
+	MatchExpressions []NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// A node selector requirement is a selector that contains values, a key, and an operator
+// that relates the key and values.
+type NodeSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators are In, NotIn, Exists, DoesNotExist. Gt, and Lt.
+	Operator NodeSelectorOperator `json:"operator"`
+	// values is an array of string values. If the operator is In or NotIn,
+	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
+	// the values array must be empty. If the operator is Gt or Lt, the values
+	// array must have a single element, which will be interpreted as an integer.
+    // This array is replaced during a strategic merge patch.
+	Values []string `json:"values,omitempty"`
```

Comment:
It's a bit of ugly that we should decide on globally.  If users are
essentially forced to write an AST wherein the describe a numeric operation
with a string value, maybe we can do better - but its not clear to me, and
I think it warrants discussion.  IntOrString is "cute" and human-friendly,
but not very friendly to automatic schema descriptions.


On Mon, Dec 28, 2015 at 3:55 PM, David Oppenheimer <notifications@github.com
> wrote:

> In docs/proposals/nodeselector.md
> <https://github.com/kubernetes/kubernetes/pull/18261#discussion_r48513984>
> :
>
> > +}
> > +
> > +// A node selector requirement is a selector that contains values, a key, and an operator
> > +// that relates the key and values.
> > +type NodeSelectorRequirement struct {
> > +	// key is the label key that the selector applies to.
> > +	Key string `json:"key" patchStrategy:"merge" patchMergeKey:"key"`
> > +	// operator represents a key's relationship to a set of values.
> > +	// Valid operators are In, NotIn, Exists, DoesNotExist. Gt, and Lt.
> > +	Operator NodeSelectorOperator `json:"operator"`
> > +	// values is an array of string values. If the operator is In or NotIn,
> > +	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
> > +	// the values array must be empty. If the operator is Gt or Lt, the values
> > +	// array must have a single element, which will be interpreted as an integer.
> > +    // This array is replaced during a strategic merge patch.
> > +	Values []string `json:"values,omitempty"`
>
> we use IntOrString as a type elsewhere in the API. Does that make this
> better or worse?
>
> I copied this from
>
> https://github.com/kubernetes/kubernetes/blob/1752cf22d4bca2fcc528042c4ab1dc2ce0947165/pkg/apis/extensions/v1beta1/types.go#L715
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/18261/files#r48513984>.
>